paddle/fluid/operators/diagonal_op.cu

```diff
@@ -0,0 +1,362 @@
+    const int64_t offset_ = context.Attr<int>("offset");
+    const int64_t axis1 = context.Attr<int>("axis1");
+    int64_t axis1_ = axis1 < 0 ? input_dim_size + axis1 : axis1;
+    const int64_t axis2 = context.Attr<int>("axis2");
+    int64_t axis2_ = axis2 < 0 ? input_dim_size + axis2 : axis2;
+    int64_t numel = input->numel();
+
+    int threads = PADDLE_CUDA_NUM_THREADS;
+    int blocks = (numel + threads - 1) / threads;
+
+    switch (input_dim_size) {
+      case 2:
+        Diagonal<T, 2, 1><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 3:
+        Diagonal<T, 3, 2><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 4:
+        Diagonal<T, 4, 3><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 5:
+        Diagonal<T, 5, 4><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 6:
+        Diagonal<T, 6, 5><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 7:
+        Diagonal<T, 7, 6><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 8:
+        Diagonal<T, 8, 7><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      case 9:
+        Diagonal<T, 9, 8><<<blocks, threads>>>(input_data, output_data, offset_,
+                                               axis1_, axis2_, input_stride,
+                                               output_stride, numel);
+        break;
+      default:
+        PADDLE_THROW(platform::errors::InvalidArgument(
+            "The rank of input should be less than 10, but received %d.",
+            input_dim_size));
+    }
+    cudaFree(input_stride);
+    cudaFree(output_stride);
+    free(host_input_stride);
+    free(host_output_stride);
+  }
+};
+
+template <typename T, int DX_DIM_SIZE, int DOUT_DIM_SIZE>
+__global__ void GradDiagonal(const T* dout_data, T* dx_data,
```

Comment:
Most of the code in Diagonal/GradDiagonal are common and can reuse.

Reply:
Done, thanks!